File: server/src/lib/embeddings/factory.ts
/**
 * Embedding Provider Factory
 * Auto-detects and creates the best available embedding provider
 *
 * Priority: Config file > Environment variables > Defaults
 */

import type { EmbeddingProvider, EmbeddingConfig } from '../../types/index.js';
import { OllamaEmbeddingProvider } from './ollama.js';
import { TransformersEmbeddingProvider } from './transformers.js';
import { HuggingFaceEmbeddingProvider } from './huggingface.js';
import { getEmbeddingSettings } from '../../utils/config.js';
import type { EmbeddingSettings } from '../../types/index.js';

/**
 * Resolve configuration from config file and environment variables
 * Config file takes precedence over env vars
 */
function resolveConfig(): EmbeddingSettings {
  const fileConfig = getEmbeddingSettings() || {};

  return {
    enabled: fileConfig.enabled,
    provider: fileConfig.provider || process.env.SAVECONTEXT_EMBEDDING_PROVIDER as EmbeddingSettings['provider'],
    HF_TOKEN: fileConfig.HF_TOKEN || process.env.HF_TOKEN,
    HF_MODEL: fileConfig.HF_MODEL || process.env.HF_MODEL,
    HF_ENDPOINT: fileConfig.HF_ENDPOINT || process.env.HF_ENDPOINT,
    OLLAMA_ENDPOINT: fileConfig.OLLAMA_ENDPOINT || process.env.OLLAMA_ENDPOINT,
    OLLAMA_MODEL: fileConfig.OLLAMA_MODEL || process.env.OLLAMA_MODEL,
    TRANSFORMERS_MODEL: fileConfig.TRANSFORMERS_MODEL || process.env.TRANSFORMERS_MODEL,
  };
}

/**
 * Detect which embedding providers are available
 * Returns providers in order of preference
 */
export async function detectAvailableProvider(): Promise<{
  available: string[];
  recommended: string | null;
}> {
  const available: string[] = [];
  const config = resolveConfig();

  // Check Ollama first (preferred for performance, local)
  const ollama = new OllamaEmbeddingProvider({
    endpoint: config.OLLAMA_ENDPOINT,
    model: config.OLLAMA_MODEL,
  });
  if (await ollama.isAvailable()) {
    available.push('ollama');
  }

  // Check HuggingFace (cloud option with custom models)
  const huggingface = new HuggingFaceEmbeddingProvider({
    apiToken: config.HF_TOKEN,
    model: config.HF_MODEL,
    endpoint: config.HF_ENDPOINT,
  });
  if (await huggingface.isAvailable()) {
    available.push('huggingface');
  }

  // Check Transformers.js (fallback, always works if installed)
  const transformers = new TransformersEmbeddingProvider({
    model: config.TRANSFORMERS_MODEL,
  });
  if (await transformers.isAvailable()) {
    available.push('transformers');
  }

  return {
    available,
    recommended: available[0] || null,
  };
}

/**
 * Create an embedding provider based on configuration or auto-detection
 * Returns null if no provider is available
 */
export async function createEmbeddingProvider(
  explicitConfig?: EmbeddingConfig
): Promise<EmbeddingProvider | null> {
  const config = resolveConfig();

  // Check if embeddings are disabled
  if (config.enabled === false) {
    return null;
  }

  // Explicit config overrides resolved config
  const provider = explicitConfig?.provider || config.provider;

  if (provider === 'ollama') {
    const ollamaProvider = new OllamaEmbeddingProvider({
      endpoint: explicitConfig?.ollamaEndpoint || config.OLLAMA_ENDPOINT,
      model: explicitConfig?.ollamaModel || config.OLLAMA_MODEL,
    });

    if (await ollamaProvider.isAvailable()) {
      return ollamaProvider;
    }

    console.warn('[SaveContext] Ollama provider configured but not available');
    return null;
  }

  if (provider === 'transformers') {
    const transformersProvider = new TransformersEmbeddingProvider({
      model: explicitConfig?.transformersModel || config.TRANSFORMERS_MODEL,
    });

    if (await transformersProvider.isAvailable()) {
      return transformersProvider;
    }

    console.warn('[SaveContext] Transformers provider configured but not available');
    return null;
  }

  if (provider === 'huggingface') {
    const hfProvider = new HuggingFaceEmbeddingProvider({
      apiToken: explicitConfig?.huggingfaceToken || config.HF_TOKEN,
      model: explicitConfig?.huggingfaceModel || config.HF_MODEL,
      endpoint: config.HF_ENDPOINT,
    });

    if (await hfProvider.isAvailable()) {
      return hfProvider;
    }

    console.warn('[SaveContext] HuggingFace provider configured but not available');
    return null;
  }

  // Auto-detect: try Ollama first, then Transformers
  const ollama = new OllamaEmbeddingProvider({
    endpoint: config.OLLAMA_ENDPOINT,
    model: config.OLLAMA_MODEL,
  });

  if (await ollama.isAvailable()) {
    return ollama;
  }

  const transformers = new TransformersEmbeddingProvider({
    model: config.TRANSFORMERS_MODEL,
  });

  if (await transformers.isAvailable()) {
    return transformers;
  }

  return null;
}

/**
 * Get embedding provider info for status display
 */
export function getProviderInfo(provider: EmbeddingProvider | null): {
  enabled: boolean;
  provider: string | null;
  model: string | null;
  dimensions: number | null;
} {
  if (!provider) {
    return {
      enabled: false,
      provider: null,
      model: null,
      dimensions: null,
    };
  }

  return {
    enabled: true,
    provider: provider.name,
    model: provider.model,
    dimensions: provider.dimensions,
  };
}

File: cli/src/embeddings/factory.rs
//! Embedding provider factory.
//!
//! Handles provider detection and creation.

use super::config::{get_embedding_settings, is_embeddings_enabled};
use super::huggingface::HuggingFaceProvider;
use super::ollama::OllamaProvider;
use super::provider::{BoxedProvider, EmbeddingProvider};
use super::types::EmbeddingProviderType;

/// Available provider detection result.
#[derive(Debug, Clone)]
pub struct ProviderDetection {
    /// List of available provider names.
    pub available: Vec<String>,
    /// Recommended provider (first available).
    pub recommended: Option<String>,
}

/// Detect which embedding providers are available.
pub async fn detect_available_providers() -> ProviderDetection {
    let mut available = Vec::new();

    // Check Ollama
    let ollama = OllamaProvider::new();
    if ollama.is_available().await {
        available.push("ollama".to_string());
    }

    // Check HuggingFace (available if token is set)
    if let Some(hf) = HuggingFaceProvider::new() {
        if hf.is_available().await {
            available.push("huggingface".to_string());
        }
    }

    // Note: We don't support transformers.js in Rust CLI
    // Users who need local inference without Ollama should use the TypeScript CLI

    let recommended = available.first().cloned();

    ProviderDetection {
        available,
        recommended,
    }
}

/// Create an embedding provider based on configuration.
///
/// Priority:
/// 1. Explicit provider in config
/// 2. Auto-detect available provider (Ollama preferred)
///
/// Returns `None` if no provider is available or embeddings are disabled.
pub async fn create_embedding_provider() -> Option<BoxedProvider> {
    // Check if embeddings are enabled
    if !is_embeddings_enabled() {
        return None;
    }

    // Check for explicit provider in config
    if let Ok(Some(settings)) = get_embedding_settings() {
        if let Some(provider_type) = settings.provider {
            return create_provider_by_type(provider_type).await;
        }
    }

    // Auto-detect: try Ollama first, then HuggingFace
    let ollama = OllamaProvider::new();
    if ollama.is_available().await {
        return Some(BoxedProvider::new(ollama));
    }

    if let Some(hf) = HuggingFaceProvider::new() {
        if hf.is_available().await {
            return Some(BoxedProvider::new(hf));
        }
    }

    None
}

/// Create a specific provider by type.
async fn create_provider_by_type(provider_type: EmbeddingProviderType) -> Option<BoxedProvider> {
    match provider_type {
        EmbeddingProviderType::Ollama => {
            let provider = OllamaProvider::new();
            if provider.is_available().await {
                Some(BoxedProvider::new(provider))
            } else {
                None
            }
        }
        EmbeddingProviderType::Huggingface => {
            HuggingFaceProvider::new().map(BoxedProvider::new)
        }
        EmbeddingProviderType::Transformers => {
            // Transformers.js is not supported in Rust CLI
            // Users should use Ollama or HuggingFace instead
            None
        }
        EmbeddingProviderType::Model2vec => {
            // Model2Vec is handled separately as fast tier
            // Use create_model2vec_provider() directly
            super::model2vec::Model2VecProvider::try_new().map(BoxedProvider::new)
        }
    }
}

/// Create a provider with explicit configuration (for testing or CLI overrides).
pub fn create_ollama_provider(endpoint: Option<String>, model: Option<String>) -> BoxedProvider {
    BoxedProvider::new(OllamaProvider::with_config(endpoint, model))
}

/// Create a HuggingFace provider with explicit configuration.
pub fn create_huggingface_provider(
    endpoint: Option<String>,
    model: Option<String>,
    token: Option<String>,
) -> Option<BoxedProvider> {
    HuggingFaceProvider::with_config(endpoint, model, token)
        .map(BoxedProvider::new)
}
